invenio_records_resources/tasks.py

```diff
@@ -1,21 +1,42 @@
 # -*- coding: utf-8 -*-
 #
-# Copyright (C) 2021 CERN.
+# Copyright (C) 2021-2022 CERN.
 #
 # Invenio-Records-Resources is free software; you can redistribute it and/or
 # modify it under the terms of the MIT License; see LICENSE file for more
 # details.
 
 """Celery tasks for async processing."""
 
+import arrow
 from celery import shared_task
+from flask import current_app
 from invenio_access.permissions import system_identity
 
-from .proxies import current_service_registry
+from .proxies import current_backrelations_registry, current_service_registry
 
 
 @shared_task(ignore_result=True)
 def extract_file_metadata(service_id, record_id, file_key):
     """Process file."""
     service = current_service_registry.get(service_id)
     service.extract_file_metadata(system_identity, record_id, file_key)
+
+
+@shared_task(ignore_result=True)
+def propagate_relation_updates(record_type, record_ids):
+    """Execute the handlers set up for a record_type update."""
+    task_start = arrow.utcnow().strftime("%Y-%m-%dT%H:%M:%S.%f")
+
+    service = current_service_registry.get(record_type)
```

Comment:
I think we might be able to avoid this read. 

1. Update runs task async by sending a list of user ids

The task will:
2. Read all changed users
3. Call each registered handler with the list of users.
4 A handler will use only the ids and revision ids of the users to search for related records that needs update.
5 The handler will then send each of the related record ids for bulk indexing.

...

6. Bulk indexing is happening at a later time and running it will read the user again.

Hence, if we look at a "change notification" instead of a list of ids, but list of (id,revision_id) pairs then we can avoid reading the user twice.

The trade-off is then when the propagation task is running, we potentially don't have the latest revision_id because more updates may have happened in-between step 1 and 2. However, since, since step 4 also takes the "indexed_at" time into account I think we'll on average have less select queries on the database.

Reply:
While I agree with this, the problem of the revision_id is bigger. When updating the task will be registered by the component (before the `RecordCommitOp`). Therefore the `revision_id` that would be passed would be outdated from the beginning.

- We can play with "`+1`" but is a hardcoded value that can be changed for some reason. It very easily leads to inconsistent data.
- It is true that having the `indexed_at`, we can simply "remove" the `@v` from the query. Worst case we would reindex records that got indexed in between the record.commit() and the task start (when it calculates the `arrow.utcnow`).

Reply:
You can probably make it work by making new uow operation:

You would pass the list of records instead of a list of id/revision_id pairs:

```python
self.uow.register(ChangeNotificationOp(record_type, [record]))
```

The task itself would the on post_commit send the celery tasks with a list of id/revision pairs (i.e. after the record has been committed and revision id bumped):

```python
class ChangeNotificationOp(Operation):
    def __init__(self, record_type, records):
        self._record_type = record_type
        self._records = records

    def on_post_commit(self, uow):
        send_change_notifications.delay(
             record_type,
             [(r.id, r.revision_id) for r in records]
        )
```

Reply:
I get the feeling of doing a very ad-hoc implementation. While I think this solves the problem of the `revision_id`, it can be that other notification handlers need some other data of the record. Then the question is: "is there such a use case?" if yes, then we might have to refactor this again, if not then I am not sure about generalizing to a "notification change system" instead of "relations update".

I'm just putting out there my thoughts, but I will implement it this way so we can move ahead :)